File: frontend/src/types/graph.ts
export interface Vulnerability {
  osvId: string;
  cveId: string | null;
  summary: string;
  cvssScore: number | null;
  severity: string | null;
  affectedRanges: string[];
}

export interface GraphNode {
  id: string;
  name: string;
  version: string;
  isRoot: boolean;
  isDirect: boolean;
  depth: number;
  vulnerabilities: Vulnerability[];
  inDegree: number;
  outDegree: number;
  betweenness: number;
  closeness: number;
  reachableFromCount: number;
}

export interface GraphEdge {
  source: string;
  target: string;
  requestedRange: string | null;
}

export interface GraphPayload {
  nodes: GraphNode[];
  edges: GraphEdge[];
}

export interface RiskScore {
  nodeId: string;
  score: number;
  rationale: string;
  contributingCves: string[];
}

export interface AnalyzeStats {
  nodeCount: number;
  edgeCount: number;
  directDependencyCount: number;
  maxDepth: number;
  vulnerableNodeCount: number;
  totalVulnerabilities: number;
  highestRiskScore: number;
}

export interface AnalyzeResponse {
  rootId: string;
  graph: GraphPayload;
  riskRanking: RiskScore[];
  stats: AnalyzeStats;
}

export interface Preset {
  id: string;
  title: string;
  year: number;
  description: string;
  package: { name: string; version: string };
}

export type AnalysisSource =
  | { kind: 'package'; name: string; version: string | null }
  | { kind: 'lockfile'; filename: string }
  | { kind: 'preset'; preset: Preset };

File: backend/app/models/graph.py
from pydantic import BaseModel

class Vulnerability(BaseModel):
    osv_id: str
    cve_id: str | None = None
    summary: str = ""
    cvss_score: float | None = None
    severity: str | None = None
    affected_ranges: list[str] = []

class GraphNode(BaseModel):
    id: str
    name: str
    version: str
    is_root: bool = False
    is_direct: bool = False
    depth: int = 0
    vulnerabilities: list[Vulnerability] = []
    in_degree: int = 0
    out_degree: int = 0
    betweenness: float = 0.0
    closeness: float = 0.0
    reachable_from_count: int = 0

class GraphEdge(BaseModel):
    source: str
    target: str
    requested_range: str | None = None

class GraphPayload(BaseModel):
    nodes: list[GraphNode]
    edges: list[GraphEdge]

class RiskScore(BaseModel):
    node_id: str
    score: float
    rationale: str
    contributing_cves: list[str] = []